lib/ide/ide/src/view/view_layout.rs

```diff
@@ -0,0 +1,151 @@
+use basegl::display::world::World;
+
+use nalgebra::zero;
+use nalgebra::Vector2;
+use std::rc::Rc;
+use std::cell::RefCell;
+use js_sys::Function;
+use web_sys::KeyboardEvent;
+use wasm_bindgen::JsCast;
+
+
+
+//TODO: ViewMode is a temporary enumeration, it will be replaced by proper Panel impl.
+// ================
+// === ViewMode ===
+// ================
+
+#[derive(Clone,Copy,Debug)]
+enum ViewMode {
+    Full,
+    Half
+}
+
+
+
+// ======================
+// === ViewLayoutData ===
+// ======================
+
+#[derive(Debug)]
+struct ViewLayoutData {
+    text_editor      : TextEditor,
+    keyboard_closure : Option<Closure<dyn FnMut(KeyboardEvent)>>,
+    view_mode        : ViewMode,
+    dimensions       : Vector2<f32>
+}
+
+impl Drop for ViewLayoutData {
+    fn drop(&mut self) {
+        self.keyboard_closure.as_ref().map(|keyboard_closure| {
+            let body = document().unwrap().body().unwrap();
+            let callback : &Function = keyboard_closure.as_ref().unchecked_ref();
+            body.remove_event_listener_with_callback("keydown", callback).ok();
+        });
+    }
+}
+
+impl ViewLayoutData {
+    fn set_view_mode(&mut self, view_mode:ViewMode) {
+        self.view_mode = view_mode;
+        self.update();
+    }
+
+    fn switch_mode(&mut self) {
+        if let ViewMode::Half = self.view_mode {
+            self.set_view_mode(ViewMode::Full)
+        } else {
+            self.set_view_mode(ViewMode::Half)
+        }
+    }
+
+    fn set_dimensions(&mut self, dimensions:Vector2<f32>) {
+        self.dimensions = dimensions;
+        self.update();
+    }
+
+    fn update(&mut self) {
+        let dimensions = self.dimensions;
+        let (position,dimensions) = match self.view_mode {
+            ViewMode::Full => {
+                let position   = Vector2::new(0.0, dimensions.y);
+                (position,dimensions)
+            },
+            ViewMode::Half => {
+                let position   = Vector2::new(0.0, dimensions.y / 2.0);
+                let dimensions = Vector2::new(dimensions.x, dimensions.y / 2.0);
+                (position,dimensions)
+            }
+        };
+        self.text_editor.set_dimensions(dimensions);
+        self.text_editor.set_position(position);
+        self.text_editor.update();
+    }
+}
+
+
+
+// ==================
+// === ViewLayout ===
+// ==================
+
+/// Initial implementation of ViewLayout with a single TextEditor. Pressing ctrl+f toggles
+/// fullscreen mode.
+#[derive(Debug,Clone)]
+pub struct ViewLayout {
+    data : Rc<RefCell<ViewLayoutData>>
+}
+
+impl ViewLayout {
+    /// Creates a new ViewLayout with a single TextEditor.
+    pub fn default(world:&World) -> Self {
+        let text_editor      = TextEditor::new(&world);
+        let keyboard_closure = None;
+        let view_mode        = ViewMode::Half;
+        let dimensions       = zero();
+        let data             = ViewLayoutData{text_editor,keyboard_closure,view_mode,dimensions};
```

Comment:
From what I see, usually we place space between the struct name and initializers opening brace.